### ISSUE TITLE
Mserver segfaults during recursive module loading

### ISSUE THREAD

**Issue by monetdb-team** [MonetDB/MonetDB#1272]
Date: 2007-03-13 02:40:35 +0100
From: Jan Rittinger &lt;<jan.rittinger>&gt;
To:   Sjoerd Mullender &lt;<sjoerd>&gt;
Version: -- development

Last updated: 2007-03-15 05:15:02 +0100

## Comment 7119
Date: 2007-03-13 14:40:35 +0100
From: Jan Rittinger &lt;<jan.rittinger>&gt;

I started my Mserver with the modules pathfinder and pftijah:
 
 Mserver --dbinit="module(pathfinder); module(pftijah); mil_start();"
 
 and tried to pose a stupid query using the MapiClient
 
 echo 1 | MapiClient -lx
 
 This leads to a more or less clean segmentation fault of the Mserver. (Note that removing 'module(pftijah);' from the dbinit string *fixes* the problem.)
 
 Looking at it with gdb indicates that the modules pathfinder and pftijah load each other recursively. I think a segfault at least isn't the expected behavior.

## Comment 7120
Date: 2007-03-15 09:12:49 +0100
From: Jan Rittinger &lt;<jan.rittinger>&gt;

Logged In: YES 
 user_id=993208
 Originator: YES
 
 The same happens if I generate a MIL script with pathfinder (e.g. echo 1 | pf > foo) and then pipe it to the Mserver (cat foo | Mserver).
 
 Calling Mserver with dbinit works:
 cat foo | Mserver --dbinit="module(pathfinder);"

## Comment 7121
Date: 2007-03-15 11:32:05 +0100
From: Sjoerd Mullender &lt;<sjoerd>&gt;

Logged In: YES 
 user_id=43607
 Originator: NO
 
 Even simpler:
 
 11:19:34+ Mserver
  Monet Database Server V4.17.1
  Copyright (c) 1993-2007, CWI. All rights reserved.
  Compiled for x86_64-redhat-linux-gnu/64bit with 64bit OIDs; dynamically linked.
  Visit http://monetdb.cwi.nl/ for further information.
 MonetDB>module(pathfinder);
  XRPC administrative console at http://localhost:50001/admin
 MonetDB>module(pftijah);
 
 at which point the server rapidly consumes infinite amounts of CPU and memory.
 Attaching gdb at this point shows that TBL_loadmod is called to load pftijah, which causes a TBL_loadmod for pathfinder, which causes a TBL_loadmod for pftijah, which causes ...
 

## Comment 7122
Date: 2007-03-15 14:19:22 +0100
From: Sjoerd Mullender &lt;<sjoerd>&gt;

Logged In: YES 
 user_id=43607
 Originator: NO
 
 The problem lies in TBL_loadmod.
 When the explicit module(pftijah) is executed, TBL_loadmod sees that the module is already loaded (because it was loaded by pathfinder).  TBL_loadmod then loads all modules on which pftijah depends, which includes pathfinder because of the .USE pathfinder statement.  So pathfinder is loaded by TBL_loadmod.  As pathfinder is also already loaded, the modules it depends on are also loaded.  In this case, that also includes all modules that were loaded by pathfinder through module statements in the pathfinder code.  One of those is pftijah.  So pftijah is loaded by TBL_loadmod.  This recurses indefinitely.
 
 Assigning to Peter.

## Comment 7123
Date: 2007-03-15 15:53:14 +0100
From: Jan Flokstra &lt;<jflokstra>&gt;

Logged In: YES 
 user_id=1054297
 Originator: NO
 
 We in Twente are also suffering from this bug. I read the comments and was a bit surprised to see a module(pftijah); in pathfinder.mx. I always thought we build pftijah 'on' Pathfinder and there were no mutual dependencies (yet?). Who made this change and why???
 A quick fix here is to remove the .USE pathfinder line in pftijah.mx and to load pathfinder every time pftijah. Peter already did this this week but removed it on my request because when you loaded pftijah without loading pathfinder first you got a very unhelpfull error message (! can't help you here, sorry.:-)
 

## Comment 7124
Date: 2007-03-15 17:15:02 +0100
From: Sjoerd Mullender &lt;<sjoerd>&gt;

Logged In: YES 
 user_id=43607
 Originator: NO
 
 Fixed by removing the circular dependency between pathfinder and pftijah.  Just always load pathfinder.  However, when you do try to load pftijah in addition to pathfinder, it now works, and if you load pftijah instead of pathfinder, you get a slightly better error message (in addition to a cryptic syntax error that we can't do much about without major effort).
 
 Note that the module(pftijah) in pathfinder.mx is necessary: pf generated code (and the code generated internally when you use MapiClient -lx) depends on pftijah, and pftijah must be loaded before mapi.
 
 In conclusion: just use pathfinder for all your XQuery needs (including XRPC and PF/TIJAH).

## Comment 7125
Date: 2010-05-04 09:32:09 +0200
From: Pseudo user for Sourceforge import &lt;<sfimport>&gt;

This bug was previously known as tracker item 1679800 at http://sourceforge.net/support/tracker.php?aid=1679800
